File: src/main/scala/tool/FirrtlCompat.scala
package tool

import firrtl.ir.Circuit
import scala.annotation.nowarn

object FirrtlCompat {
  @nowarn("cat=deprecation")
  def parseCircuit(text: String): Circuit = {
    try {
      val parserClass = Class.forName("firrtl.Parser$")
      val module = parserClass.getField("MODULE$").get(null)
      parserClass.getMethod("parse", classOf[String]).invoke(module, text).asInstanceOf[Circuit]
    } catch {
      case e: ReflectiveOperationException =>
        throw new RuntimeException("Failed to invoke FIRRTL parser", e)
    }
  }
}

File: src/main/java/tool/FirrtlCompat.java
package tool;

import firrtl.ir.Circuit;

public final class FirrtlCompat {
  private FirrtlCompat() {}

  public static Circuit parseCircuit(String text) {
    try {
      Class<?> parserClass = Class.forName("firrtl.Parser$");
      Object module = parserClass.getField("MODULE$").get(null);
      return (Circuit) parserClass.getMethod("parse", String.class).invoke(module, text);
    } catch (ReflectiveOperationException e) {
      throw new RuntimeException("Failed to invoke FIRRTL parser", e);
    }
  }
}
